keps/sig-scheduling/4328-node-affinity-based-eviction/README.md

```diff
@@ -0,0 +1,459 @@
+
+Without `node-affinity-eviction`, I have to relabel the node to suggest data on that node is corrupted, 
+and then manually iterating the pods on the node, decide whether to delete them to trigger a reschedule.
+With `node-affinity-eviction`, the only thing I need to do is to relabel the node, and `node-affinity-eviction` controller will do the rest.
+
+### Notes/Constraints/Caveats (Optional)
+
+N/A
+
+### Risks and Mitigations
+
+The major concern for this controller may be the overhead of the nodes with a large number of pods. 
+Every time a node's label is updated, the controller needs to iterate all pods assigned to the node. This may cause potential scaling issues.
+However, the impact won't be significant due to the limitation on the number of pods per node.
+The evictions caused by this may be expensive depending on the number of pods.
+
+## Design Details
+
+Introducing a new type of NodeAffinity:
+```go
+type NodeSelector struct {
+	NodeSelectorTerms []NodeSelectorTerm
+}
+
+type NodeAffinity struct {
+	// If the affinity requirements specified by this field are not met at
+	// scheduling time, the pod will not be scheduled onto the node.
+	// If the affinity requirements specified by this field cease to be met
+	// at some point during pod execution (e.g. due to an update), the system
+	// will try to eventually evict the pod from its node.
+	// +optional
+	RequiredDuringSchedulingRequiredDuringExecution *NodeSelector
+}
+```
+
+Add a controller called `node-affinity-eviction` 
+
+The controller do the following things:
+
+- Listening to the changes of node labels
+- Iterating over all pods assigned to the node(excluding mirror pods), checks the NodeAffinity field, if `RequiredDuringSchedulingRequiredDuringExecution` exists, checks if `NodeSelector` still match the new node. 
+- If `RequiredDuringSchedulingRequiredDuringExecution` is no loger met, trying to evict the pod.
+- It will use Eviction API to evict the pod(The API takes the PDBs into consideration , so we do not have to use the PDBs explicitly here). 
+- If eviction fails, we will take different actions based on the failure reasons:
+  - If eviction fails due to a violation of the PodDisruptionBudget (PDB), we will keep the original Pod unchanged and ignore the eviction. 
+  - If eviction fails for any other reason, we will attempt a certain number of retries. If the eviction continues to fail after multiple retries, we will keep the original Pod unchanged and ignore the eviction. 
+  - Relevant logs will be printed for every failure.
+
+During the alpha stage, we will use a simple eviction API to evict Pods. This may result in Pods that fail eviction needing a change in the node's labels to trigger rescheduling again.
+
+During the beta stage, we will switch to using the [Evacuation API](https://github.com/kubernetes/enhancements/issues/4563) if the user enables the corresponding feature gate. This will address a series of issues related to failed evictions.
+
```

Comment:
the daemonset parts should be mentioned also in the implementation, since it requires changes to the controller

Reply:
> since it requires changes to the controller

You mean daemonset controller?

I'm not familiar with ds controller. But I think ds controller will handle node label updates also?

> If node labels are changed, the DaemonSet will promptly add Pods to newly matching nodes and delete Pods from newly not-matching nodes.

For ds controller, this affinity should look like existing affinity: `requiredDuringSchedulingIgnoredDuringExecution`, and I can't think of which part of the ds controller we should modify.

Reply:
ds controller does its own scheduling so it requires code changes there for the de/scheduling bits